### ISSUE TITLE
How to add custom model into LIT?

### ISSUE THREAD

**Issue by abhibisht89** [PAIR-code/lit#599]
how to add your own NER model and dataset into lit.

Any help ?

Thanks

**Comment by jameswex** [comment]
See https://github.com/PAIR-code/lit/wiki/api.md#adding-models-and-data for defining your own model and dataset for use in LIT. You just need to write a Model and Dataset wrapper for your model and dataset. 

You can see and run code for a demo of the Stanza NER model in LIT here: https://github.com/PAIR-code/lit/blob/main/lit_nlp/examples/stanza_demo.py, with the model defined at https://github.com/PAIR-code/lit/blob/main/lit_nlp/examples/models/stanza_models.py

**Comment by pratikchhapolika** [comment]
Hi @jameswex any end to end code to add custom model and dataset for Roberta classification task. Or, are you planning to have one notebook(.ipynb) to get started with custom models and datasets. For beginners like me its little confusing. 
Apologies!

**Comment by SpamDude15** [comment]
How would I add an image classification model? I have a `.h5` file of the model, but I'm having a hard time loading it.

**Comment by jameswex** [comment]
@SpamDude15, as per https://github.com/PAIR-code/lit/issues/599, you should create a class for your model that inherits from `Model`. Inside that class, you can load up your model in whichever way you like in the class constructor, and then define the `predict` method to use that model to make predictions on the inputs passed to it and return the results. You will also have to define the `input_spec` and `output_spec` methods on the class to specify what the inputs to the model are and what the outputs from the model are. 

One image model example can be found at https://github.com/PAIR-code/lit/blob/main/lit_nlp/examples/models/mobilenet.py

**Comment by RyanMullins** [comment]
> Hi @jameswex any end to end code to add custom model and dataset for Roberta classification task. Or, are you planning to have one notebook(.ipynb) to get started with custom models and datasets. For beginners like me its little confusing. Apologies!

@pratikchhapolika The LIT team does not have time to implement base classes for all models. If you're interested in RoBERTa, you can access the model from [Hugging Face](https://huggingface.co/roberta-base) and [implement your own `Model` class](https://github.com/PAIR-code/lit/wiki/api.md#adding-models-and-data). The key things to implement 

As for data, LIT provides 2 of the 4 datasets ([MNLI](https://github.com/PAIR-code/lit/blob/main/lit_nlp/examples/datasets/glue.py#L258) and [SST-2](https://github.com/PAIR-code/lit/blob/main/lit_nlp/examples/datasets/glue.py#L55)) used in the [RoBERTa paper](https://arxiv.org/pdf/1907.11692.pdf) out of the box. We don't have wrappers for SQuAD (on [TFDS](https://www.tensorflow.org/datasets/catalog/squad) or [HF Datasets](https://huggingface.co/roberta-base)) or RACE (on [TFDS](https://www.tensorflow.org/datasets/catalog/race) or [HF Datasets](https://huggingface.co/datasets/race)) at this time. 

Once you have the `Model` and `Dataset` classes of interest, it's relatively simple to initialize a LIT app instance either as a [Python module](https://github.com/PAIR-code/lit/blob/main/lit_nlp/examples/penguin_demo.py#L68) (note I referenced our Penguin demo because it's among the simplest of the built-ins) or [in Colab](https://colab.research.google.com/github/PAIR-code/lit/blob/main/lit_nlp/examples/notebooks/LIT_sentiment_classifier.ipynb).



For all commenters on this issue (@pratikchhapolika @SpamDude15 @abhibisht89), please remember that if you do implement these classes or demo modules, we would welcome [contributions via PRs](https://github.com/PAIR-code/lit/blob/main/README.md#pull-request-process) to incorporate them into the core LIT example.

**Comment by SpamDude15** [comment]
@jameswex I was wondering if there was a dynamic option to load models and datasets into LIT without having to write three different python files for loading, model, and dataset.

**Comment by RyanMullins** [comment]
@SpamDude15 not for models. There are too many modeling frameworks, and increasingly serving frameworks, for our research team to keep up with. Instead we've chosen to go the self-service route. 

Datasets are slightly easier. We are working on a generic dataset wrapper for TFDS, though I don't have an ETA for its release. That will likely act as a reference for creating a generic wrapper for the HF Datasets library, too, but again, no ETA on release. 

There is a [Blank Slate demo](https://github.com/PAIR-code/lit/blob/main/lit_nlp/examples/blank_slate_demo.py) that's easily forkable. Once you have the model and dataset classes, it should be as simple as importing them into the demo file and launching.

**Comment by SpamDude15** [comment]
@RyanMullins That's interesting, is there a way I can load in .ppm images? I looked through the datatypes on LIT, and there seems to be no option for ppm, so does that mean it isn't compatible?

**Comment by SpamDude15** [comment]
@RyanMullins I got PPM images to work with LIT using base64 encoding, which was great, I was wondering if PNG could be loaded in the same way without having to use the defined datatype by LIT.
This is how I load PPM images, this should work with any image type as well right?
```python
image_file_path = Path(base_path+path_name) / row['Filename']
pil_image = PILImage.open(image_file_path)
image_str = image_utils.convert_pil_to_image_str(pil_image)
```

**Comment by RyanMullins** [comment]
Yup, that'll work for pretty much any image Pillow supports. The more specific`ImageBytes` subclasses are really just there for people who do pre- or post-processing on images and want a more type-safe way of differentiating code paths.
